### PR TITLE
Introduce the "Unlit" shading model support and apply it to Web Enitities

### PR DESCRIPTION
**Pull request by samcake**
- Replace the concept of  "emissive" shading used for objects meant to be Unlit and not affected by lighting by "Unlit". this affects severall shaders and piepline key flags which are now called Unlit.
Emissive is somewhat confusing with the true emissive component that can also be applied to lit objects.

- Add support for "Unlit" material from FBX.
 Material need to have a shading model called "Lambert" and a diffuseFactor  set to 0.
The emissiveColor or emissiveTexture is then  recognized as the "unlit" color.
 In Maya: pick Lambert Material, set diffuse to 0 and the color is assigned from  the "Incandescence" property, a texture can be assigned to the incandesnce field too.
 In Blender: select Lambert for the Diffuse model, select  "Wardiso" for the specular model (generating a Lamber Material during the FBX export). Turn the Diffuse intensity to 0. Check the "Shadeless" box, turn the emit slider to be 1.0. The color is assigned from the actual diffuse color. A texture can be assigned as the unlit texture by picking the Emit influence for it.

- Use the "Unlit" shading model for the Web entities so they are not affected by the lighting environment and look like in a regular browser.
Add to hack a sRGB to linear conversion on the web entity texture created from qml because the format is not sRGB. this is done by passing an ALpha of 0 to the GeometryCache primitive calls.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 303cade54782d46613658a9e2aab97f260351e49 available at: https://gist.github.com/33c5b95cb6a9c607fcfdbcd3f80cf232<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7859.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7859.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e40a795a27f11479f0ff1623430daf90a71a22e6 available at: https://gist.github.com/33c5b95cb6a9c607fcfdbcd3f80cf232<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7859.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7859.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1ec9ef560e472462da07c30dda2a8f6f1126e122 available at: https://gist.github.com/33c5b95cb6a9c607fcfdbcd3f80cf232<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7859.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7859.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 99dcfef5a8f9eab81d2d7b5637ea66ed9d29d5ae available at: https://gist.github.com/33c5b95cb6a9c607fcfdbcd3f80cf232<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7859.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7859.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 80c80091e4f755e572e9cca3d0863fd3c181a24b available at: https://gist.github.com/33c5b95cb6a9c607fcfdbcd3f80cf232<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7859.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7859.dmg

**Comment by samcake** [comment]
To Test:
1/ Web entities should look good and the color perceived should be very close to what you see in a regular browser in 2D. There should be NO lighting effect so changing the zone key light or entities should not affect it, adding local lights should not affect it

2/ New models containing the unlit material description should be tries and you should see the difference with the current master build:
here to test 4 models, 2 for blender, 2 for maya. THey each represent the case with and with out texture.
http://hifi-public.s3.amazonaws.com/sam/models/testModels/maya/torusLambertIncandescenceRed.fbx
http://hifi-public.s3.amazonaws.com/sam/models/testModels/maya/torusLambertIncandescenceTextured.fbx

http://hifi-public.s3.amazonaws.com/sam/models/testModels/blender/torusShadelessGreen.fbx
http://hifi-public.s3.amazonaws.com/sam/models/testModels/blender/torusShadelessGreenTextured.fbx

Thanks

S

**Comment by Nex-Pro** [comment]
@samcake 

#### Looks good to me.

Models ignore local lighting effects. 

In the scene shown in the images lighting is off apart from a blue light with intensity of 4000. The blue light is illuminating the top of the grey cube. Which is mostly showing as solid black.

In one of the images the grey cube is lit with three lights. Red Green and Blue. The models are unaffected.

**It would be nice to have the unlit mode as an option in the entity properties. As it offers a simple way to make objects self-illuminate.**

The uploaded models look self-illuminated as they should.

There are some differences in the colors of the web entity as compared to my chrome browser. ( this is to be expected I feel ).


tonality_78592

rgb_lights_7859

tonality_7859

** Side note blender units are Metres Maya is centimeters. **

**Comment by Nex-Pro** [comment]
@sam it crashed on exit. Bugsplat sent.

File: tools/scribe/src/main.cpp [hunk 1]
```diff
@@ -191,10 +191,19 @@ int main (int argc, char** argv) {
         targetStringStream << "#ifndef scribe_" << targetName << "_h" << std::endl;
         targetStringStream << "#define scribe_" << targetName << "_h" << std::endl << std::endl;
 
-        // targetStringStream << "const char " << targetName << "[] = R\"XXXX(" << destStringStream.str() << ")XXXX\";";
+        std::istringstream destStringStreamAgain(destStringStream.str());
```
Comment (jherico) [inline tools/scribe/src/main.cpp]: If we're actually hitting the 65k max string literal length, then we should  break up the string into as few chunks as possible, prefereably as the last line feed before the issue.

**Comment by samcake** [comment]
@Nex-Pro I can't repro the crash on exit?

File: tools/scribe/src/main.cpp (hunk 1 above)
Comment (samcake) [inline tools/scribe/src/main.cpp]: I agree, let me fix that before merging

**Comment by hifi-gustavo** [comment]
Build console output for commit: a436c54817e3194748e1957fca46d3187c9138aa available at: https://gist.github.com/33c5b95cb6a9c607fcfdbcd3f80cf232<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7859.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7859.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 670826effc912d30c20053860e7f25c3b36090b9 available at: https://gist.github.com/33c5b95cb6a9c607fcfdbcd3f80cf232<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7859.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7859.dmg

**Comment by Nex-Pro** [comment]
@sam I just left it running for a good while. Maybe more than a half hour. Then it crashed on exit.

**Pull request merged**